gherkin/c/CMakeLists.txt

```diff
@@ -67,7 +67,11 @@ LIST(APPEND GHERKIN_CLI
 
 
 add_library(gherkin ${GHERKIN_SRS})
-target_link_libraries(gherkin PRIVATE m)
+include(CheckLibraryExists)
+CHECK_LIBRARY_EXISTS(m log10 "" HAVE_LIB_M)   
+if (HAVE_LIB_M)  
+    target_link_libraries(gherkin PRIVATE m)
+endif (HAVE_LIB_M)
```

Comment:
it is checking for the symbol log10 in the library, log10 is the function that is being used in math.h.
https://github.com/cucumber/cucumber/blob/4c14c9ac7d0bd8aaedbc7a0e8ee5d3b31756f8e2/gherkin/c/src/error_list.c#L216
I don't know what the significance of it is, just that it is being used.

Reply:
You are implying that math.h is generally available. And checking for log10 in lib m is exceptional. Correct?

Reply:
Nothing exceptional, just msvc does it different than clang and gcc.  

math.h contains the declarations for the library.  In gcc & clang the definitions for the functions in math.h are contained in m lib.  In windows the definitions are included in the normal msvcrt runtime, and there is no m library.  So if you always try to link it in you get an error, because there is no m lib.

So this additions says, if there is a m library, and it contains a definition for log10, then link it in.  If the condition fails, then assume that it is part of the normal runtime, and does not need to be specifically linked in. 

When I ran into this error, searching led me to this solution.   It seems like maybe it is a common enough issue in compiling something written in gcc/clang and later compiled with MSVC that it was the majority of the 1st page.

In looking at it, it seems that math.h is part of the C standard, and I guess probably it is left up to the compiler implementation to decide where the library implementations go.  Or these decisions were made before it was standardized.

There may be some other vendor that defines math.h (I believe they must to conform with ISO), but they decide to call it math.lib instead of m or being just built in.  The build would fail on this compiler, and the error would be a link error for unresolved external.  Meaning we tried to call log10 declared math.h, but there is no implementation of the function.  And we'd have to add another check to add support for that compiler, and link in whatever library they decided to put the math.h function definitions in.

So it would be nice if the compiler could automatically figure out that if you include math.h, do the right thing, but unfortunately C/C++ doesn't work like that.  The reason for this is that there are 3 basic steps in C/C++ compilation.

1) preprocessor - replace all preprocessor defintions.  #include <blah.h> gets replaced with the contents of blah.h.  and all of the macros get replaced with whatever they were #define.
2) compiler - each .c, .cpp (or other file you feed to the compiler) gets compiled into a separate object file.  Object files can have definitions, and a record of declarations that they need a definition to (externals).  But they can't really do anything, because they don't yet know about any of the other object files.
3) Linker - linker takes in all of the object files, and any library files you tell it, builds a graph of the definitions provided in each object file, and the definitions needed in each object file, and makes everything work, by pointing the external calls to the right definition.

Step 3 is where it is currently broken on msvc, because we told it there is an m.lib, but it doesn't exist, and so even though it does have everything it needs to complete the link, it lets us know that we told it to link in something and it can't find it.

Reply:
Cheers!